File: src/main/java/org/pixeltime/enchantmentsenhance/event/enchantment/Web_Trap.java
package org.pixeltime.enchantmentsenhance.event.enchantment;

import com.sk89q.worldguard.bukkit.WGBukkit;
import com.sk89q.worldguard.protection.flags.DefaultFlag;
import com.sk89q.worldguard.protection.flags.StateFlag;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.scheduler.BukkitRunnable;
import org.pixeltime.enchantmentsenhance.Main;
import org.pixeltime.enchantmentsenhance.manager.SettingsManager;

import java.util.ArrayList;

public class Web_Trap implements Listener {
    private ArrayList<Block> temp;

    public Web_Trap() {
        this.temp = new ArrayList<Block>();
    }

    @EventHandler(priority = EventPriority.MONITOR)
    public void onDamage(final EntityDamageByEntityEvent entityDamageByEntityEvent) {
        final String translateAlternateColorCodes = ChatColor.translateAlternateColorCodes('&', SettingsManager.lang.getString("enchantment." + "web_trap"));
        if (entityDamageByEntityEvent.getEntity() instanceof Player && entityDamageByEntityEvent.getDamager() instanceof Player) {
            try {
                final Player player = (Player) entityDamageByEntityEvent.getEntity();
                final Player player2 = (Player) entityDamageByEntityEvent.getDamager();
                if (entityDamageByEntityEvent.isCancelled()) {
                    return;
                }
                if (SettingsManager.enchant.getBoolean("allow-worldguard") && WGBukkit.getRegionManager(player.getWorld()).getApplicableRegions(player.getLocation()).queryState(null, new StateFlag[]{DefaultFlag.PVP}) == StateFlag.State.DENY) {
                    return;
                }
                final int n = (int) (Math.random() * 100.0);
                if (player2.getItemInHand().getItemMeta().getLore().contains(String.valueOf(translateAlternateColorCodes) + " I") && n < SettingsManager.enchant.getInt("web_trap.level_I.chance")) {
                    if (SettingsManager.enchant.getBoolean("web_trap.check-wg-flag") && WGBukkit.getRegionManager(player.getWorld()).getApplicableRegions(player.getLocation()).queryState(null, new StateFlag[]{DefaultFlag.BUILD}) == StateFlag.State.DENY) {
                        return;
                    }
                    final Location location = player.getLocation();
                    final Block block = location.getBlock();
                    this.temp.add(block);
                    location.getBlock().setType(Material.WEB);
                    Bukkit.getServer().getScheduler().scheduleSyncDelayedTask(Main.getMain(), new BukkitRunnable() {
                        public void run() {
                            block.setType(Material.AIR);
                            Web_Trap.this.temp.clear();
                        }
                    }, (long) (SettingsManager.enchant.getInt("web_trap.level_I.duration") * 20));
                }
                if (player2.getItemInHand().getItemMeta().getLore().contains(String.valueOf(translateAlternateColorCodes) + " II") && n < SettingsManager.enchant.getInt("web_trap.level_II.chance")) {
                    if (SettingsManager.enchant.getBoolean("web_trap.check-wg-flag") && WGBukkit.getRegionManager(player.getWorld()).getApplicableRegions(player.getLocation()).queryState(null, new StateFlag[]{DefaultFlag.BUILD}) == StateFlag.State.DENY) {
                        return;
                    }
                    final Location location2 = player.getLocation();
                    this.temp.add(location2.getBlock());
                    location2.getBlock().setType(Material.WEB);
                    Bukkit.getServer().getScheduler().scheduleSyncDelayedTask(Main.getMain(), new BukkitRunnable() {
                        public void run() {
                            location2.getBlock().setType(Material.AIR);
                            Web_Trap.this.temp.clear();
                        }
                    }, (long) (SettingsManager.enchant.getInt("web_trap.level_II.duration") * 20));
                }
                if (player2.getItemInHand().getItemMeta().getLore().contains(String.valueOf(translateAlternateColorCodes) + " III") && n < SettingsManager.enchant.getInt("web_trap.level_III.chance")) {
                    if (SettingsManager.enchant.getBoolean("web_trap.check-wg-flag") && WGBukkit.getRegionManager(player.getWorld()).getApplicableRegions(player.getLocation()).queryState(null, new StateFlag[]{DefaultFlag.BUILD}) == StateFlag.State.DENY) {
                        return;
                    }
                    final Location location3 = player.getLocation();
                    this.temp.add(location3.getBlock());
                    location3.getBlock().setType(Material.WEB);
                    Bukkit.getServer().getScheduler().scheduleSyncDelayedTask(Main.getMain(), new BukkitRunnable() {
                        public void run() {
                            location3.getBlock().setType(Material.AIR);
                            Web_Trap.this.temp.clear();
                        }
                    }, (long) (SettingsManager.enchant.getInt("web_trap.level_III.duration") * 20));
                }
            } catch (Exception ex) {
            }
        }
    }
}

File: src/main/kotlin/org/pixeltime/enchantmentsenhance/event/enchantment/Web_Trap.kt
/*
 *     Copyright (C) 2017-Present HealPotion
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package org.pixeltime.enchantmentsenhance.event.enchantment

import com.sk89q.worldguard.bukkit.WGBukkit
import com.sk89q.worldguard.protection.flags.DefaultFlag
import com.sk89q.worldguard.protection.flags.StateFlag
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.scheduler.BukkitRunnable
import org.pixeltime.enchantmentsenhance.Main
import org.pixeltime.enchantmentsenhance.manager.IM
import org.pixeltime.enchantmentsenhance.manager.KM
import org.pixeltime.enchantmentsenhance.manager.SettingsManager
import java.util.*

class Web_Trap : Listener {
    private val temp: ArrayList<Block> = ArrayList()

    @EventHandler(priority = EventPriority.MONITOR)
    fun onDamage(entityDamageByEntityEvent: EntityDamageByEntityEvent) {
        val translateAlternateColorCodes = ChatColor.translateAlternateColorCodes('&', SettingsManager.lang.getString("enchantments." + "web_trap"))
        if (entityDamageByEntityEvent.entity is Player && entityDamageByEntityEvent.damager is Player) {
            try {
                val player = entityDamageByEntityEvent.entity as Player
                val player2 = entityDamageByEntityEvent.damager as Player
                if (entityDamageByEntityEvent.isCancelled) {
                    return
                }
                if (SettingsManager.enchant.getBoolean("allow-worldguard") && WGBukkit.getRegionManager(player.world).getApplicableRegions(player.location).queryState(null, DefaultFlag.PVP) == StateFlag.State.DENY) {
                    return
                }
                val armorContents = IM.getArmorSlots(player2) + IM.getAccessorySlots(player2)
                for (itemStack in armorContents) {
                    if (itemStack.hasItemMeta() && itemStack.itemMeta.hasLore()) {
                        val level = KM.getLevel(translateAlternateColorCodes, itemStack.itemMeta.lore)
                        if (level > 0 && (Math.random() * 100.0).toInt() < SettingsManager.enchant.getInt("web_trap.$level.chance")) {
                            if (SettingsManager.enchant.getBoolean("web_trap.check-wg-flag") && WGBukkit.getRegionManager(player.world).getApplicableRegions(player.location).queryState(null, DefaultFlag.BUILD) == StateFlag.State.DENY) {
                                return
                            }
                            val location = player.location
                            val block = location.block
                            this.temp.add(block)
                            location.block.type = Material.WEB
                            Bukkit.getServer().scheduler.scheduleSyncDelayedTask(Main.getMain(), object : BukkitRunnable() {
                                override fun run() {
                                    block.type = Material.AIR
                                    this@Web_Trap.temp.clear()
                                }
                            }, (SettingsManager.enchant.getInt("web_trap.$level.duration") * 20).toLong())
                        }
                    }
                }
            } catch (ex: Exception) {
            }

        }
    }
}
